nix/t1/submodules.nix

```diff
@@ -15,27 +15,28 @@ in
     }
 
     _setupOneSubmoduleEditable() {
-      src="$1"
-      name="$2"
-      remote="$3"
+      name="$1"; shift
+      remote="$1"; shift
+      rev="$1"; shift
       depDir="dependencies/$name"
 
       if [ -d "$depDir" -a ! -L "$depDir" ]; then
         echo "[nix-shell] ignored existing submodule directory '$depDir', remove them if you want a update"
       else 
         if [ -L "$depDir" ]; then
-          echo "[nix-shell] replacing symlink '$depDir' with '$src'"
+          echo "[nix-shell] replacing symlink '$depDir' with full git worktree"
           rm "$depDir"
         else
-          echo "[nix-shell] copying '$src' to '$depDir'"
+          echo "[nix-shell] fetching submodule $name"
         fi
-        cp -rT "$src" "$depDir"
-        chmod +w -R "$depDir"
-      fi
 
-      if ! git -C "$depDir" remote get-url origin >/dev/null 2>&1; then
-        echo "[nix-shell] adding remote '$remote' for '$depDir'"
-        git -C "$depDir" remote add origin "$remote"
+        git clone $remote $depDir
+        pushd $depDir > /dev/null
+        git -c advice.detachedHead=false checkout $rev
+        popd > /dev/null
+
+        echo "[nix-shell] checkout to *detached head*: $rev for submodule $name"
```

Comment:
Logging here seems unnecessary here since git checkout would print the commit being checked out